exasol_transformers_extension/udfs/models/filling_mask_udf.py

```diff
@@ -0,0 +1,202 @@
+        """
+        result_df_list = []
+        unique_values = dataframe_operations.get_sorted_unique_values(
+            batch_df, constants.ORDERED_COLUMNS)
+        for model_name, bucketfs_conn, sub_dir in unique_values:
+            model_df = batch_df[
+                (batch_df['model_name'] == model_name) &
+                (batch_df['bucketfs_conn'] == bucketfs_conn) &
+                (batch_df['sub_dir'] == sub_dir)]
+
+            current_model_key = (bucketfs_conn, sub_dir, model_name)
+            if self.last_loaded_model_key != current_model_key:
+                self.set_cache_dir(model_df)
+                self.clear_device_memory()
+                self.load_models(model_name)
+                self.last_loaded_model_key = current_model_key
+
+            unique_top_k_values = model_df['top_k'].unique()
+            for top_k in unique_top_k_values:
+                model_by_top_k_df = model_df[model_df['top_k'] == top_k]
+                self.setup_pipeline(top_k=top_k)
+                model_pred_df = self.get_prediction(model_by_top_k_df)
+                result_df_list.append(model_pred_df)
+
+        result_df = pd.concat(result_df_list)
+        return result_df
+
+    def set_cache_dir(self, model_df: pd.DataFrame) -> None:
+        """
+        Set the cache directory in bucketfs of the specified model. Note that,
+        cache_dir class variable is used for testing purpose. This variable is
+        set to a local path only in unit tests.
+
+        :param model_df: The model dataframe to set the cache directory
+        """
+        model_name = model_df['model_name'].iloc[0]
+        bucketfs_conn_name = model_df['bucketfs_conn'].iloc[0]
+        sub_dir = model_df['sub_dir'].iloc[0]
+        bucketfs_location = bucketfs_operations.create_bucketfs_location(
+            self.exa.get_connection(bucketfs_conn_name))
+
+        model_path = bucketfs_operations.get_model_path(sub_dir, model_name)
+        self.cache_dir = bucketfs_operations.get_local_bucketfs_path(
+            bucketfs_location=bucketfs_location, model_path=str(model_path))
+
+    def load_models(self, model_name: str, **kwargs) -> None:
+        """
+        Load model and tokenizer model from the cached location in bucketfs
+
+        :param model_name: The model name to be loaded
+        """
+        self.last_loaded_model = self.base_model.from_pretrained(
+            model_name, cache_dir=self.cache_dir)
+        self.last_loaded_tokenizer = self.tokenizer.from_pretrained(
+            model_name, cache_dir=self.cache_dir)
+
+    def setup_pipeline(self, **kwargs) -> None:
+        """
+        Setup pipeline with the loaded models and the current topk value
+        """
+
+        self.last_created_pipeline = self.pipeline(
+            "fill-mask",
+            model=self.last_loaded_model,
+            tokenizer=self.last_loaded_tokenizer,
+            device=self.device,
```

Comment:
I think, we need to put the model already before creating the pipeline onto the device, if we recreate the pipeline all the time

Reply:
good point